ib2roce/man/ib2roce.1
.\" Licensed under the OpenIB.org BSD license (FreeBSD Variant) - See COPYING.md
.\"
.\" Copyright (C) 2023 Christoph Lameter <cl@linux.com>
.\"
.TH "ib2roce" 1 "2023-4-14" "ib2roce" "ib2roce" ib2roce
.SH NAME
ib2roce \- Infiniband to ROCE gateway
.SH SYNOPSIS
.sp
.nf
\fIib2roce\fR [<options>] [-d <infiniband device] [-r <roce device>] [-m <multicast address>] .. 
.fi
.SH "DESCRIPTION"
.B ib2roce
is a bridging tool that monitors and forwards RDMA packets
between an Infiniband fabric and a Ethernet fabric. On the Ethernet
side ROCE v2 is used for RDMA communication.

ib2roce is focused on forwarding multicast datagrams and supports the PGM
protocol (RFC3208) typically used to implement reliable multicast
operations. Unicast support for NAKs is there so that PGM streams
recovery can be fowarded through the bridge [experimental] but the
NAKs can also be forwarded via other mechanisms.

True RDMA unicast operations are not considered
necessary since the primary role of ib2roce is to support multicast
messaging on RDMA fabrics that need unicast only as a recovery
control channel.
.B ib2roce
is optimized to provide the highest
throughput for multicast messaging an can be configured for
low latency on NOHZ_FULL cores.
.SH "OPTIONS"
.TP
\-c, --config <file>
Read options an configuration parmaeters from a file. The file needs to use the
long form of the options. Each line contains one option.
Comment lines are prefixed by # and empty lines are ignored.
See for example the configuration file in /etc/rdma/ib2roce.cfg
.TP
\-d, --device  <infiniband-device[:port]>
Specifies the Infiniband device to use with an optional port number.
The port defaults to one if not specified.

.B ib2roce
will scan all devices on startup and will use the first
Infiniband device that is suitable. If there are
multiple Infiniband interfaces available on host then this
option is typically necessary to identify the one to use.
.TP
\-e, --enable <options>[=<value]
Set or enable an optinal feature in the ib2roce daemon. See the section on optional features for details.
Specifying this option without a feature will print a list of possible features to enable.
.TP
\-h, --help
Display a list of options that can be specified on the command line or in a configuration file
.TP
\-i, --inbound <ipv4 multicast address[:port][/<mgid-format]
A multicast address that should be forwarded from Infiniband
to ROCE. Traffic in the other direction is ignored.
.TP
\-k, --cores [<number-cores>]
Enable low latency operation on the given number of cores.
It is recommended that the system operate with NOHZ_FULL and that the
operating system threads etc are restricted to a NUMA node different
than where the Infiniband and ROCE NICs are located.
.B ib2roce
will place RDMA channels in such a way that a sending and receiving
channel is paired on each core. The number of pairs depends on the
number of RDMA channels needed to accomodate the multicast groups
being forwarded. If there are more pairs than the
number of configured cores then
.B ib2roce
will place multiple pairs on individual cores.
These special low latency cores will operate in a busy loop.
Without the
.B --cores
option all processing will be done by the high latency thread.
The high latency thread will multiplex operations for all hardware, software and timer events in a
single thread.
Without the
.B --cores
options ib2roce is a
.I single threaded process
where synchronization can be avoided through cooperative multitasking.
Switching on busy loop cores results in additional overhead through
the additional synchronization for concurrent operations.
.TP
\-l, --mgid [<mgid-format>]
Specifies which MGID format to use on Infiniband. There are a
couple of standards on how to encode ipv4 addresses in the
MGID on Infiniband. This selects
the method to use and its use is depending on the type of traffic
that needs to be forwarded between the fabrics. Note that -l
only affects the multicast groups specified on the command line
.I after
the option was specified.

If no format is specified then ib2roce will display the available
formats. If no custom mgid format is set for a multicast group
(the default) then ib2roce will perform a join using PORT conventions
on Infiniband.

The PORT MGID format relies on the use of a port number to construct
the MGID. This means that the port number needs to be correctly
configured otherwise the MGID will not match and traffic will not
flow to and from applications relying on PORT. Other MGID encoding
schemes do not include the port and there is also no way to specify
an effective port numer for UDP datagrams on the ROCE side. It is
therefore recommended to use a standard port number for all
multicast groups.

.TP
\-m, --multicast <ipv4 multicast address[:port][/<mgid-format]>
Specify a multicast address that should be forwarded in both directions.
The port specification is optional since it is often not relevant for multicast
traffic forwarding. The mgid format can be specified to override
the global setting. See the -l option for more details.
.TP
\-o, --outbound <ipv4 multicast address[:port][/<mgid-format]
A multicast address that should be forwarded from ROCE
to Infiniband. Traffic in the other direction is ignored. Sendonly joins
occur to avoid additional traffic on the fabric.
.TP
\-p, --port <number>
Set the default port number. This is used when binding to ROCE and Infiniband
devices and for multicast groups that do not have a port specified.
Currently these port numbers seem to be mostly ignored by the RDMA subsystem.
However, the port number is relevant when an PORT type MGID needs to be
constructed. The option affects all following multicast group specifications.
.TP
\-q, --test
Run verification tests for the internal hashing and queuing mechanisms.
.TP
\-r, --roce  <roce-device[:port]>
Specifies the ROCE Ethernet device to use with an optional port number.
The port defaults to one if not specified.

.B ib2roce
will scan all devices on startup and will use the first Ethernet
device that supports ROCE v2 and IPv4 by default. If there are
multiple ROCE fabrics attached then this option is usually necessary
to identify the one to use.
.TP
\-s, --systemd
Run ib2roce from systemd. Do not create a pid file and use the systemd libraries
for logging and process state notifications.
.TP
\-t, --tos <value>
Set a TOS value. This is set via
.B rdma_set_options
when an RDMA channel is opened and is also set as the traffic class when sending.
This allows setup of the DSCP bits in the IP header on ROCE and the traffic class
for the serice level determination on Infiniband.
The option affects all multicast groups specified
.I after
this option was given.
.TP
\-v, --verbose
Increase verbosity for the logging level. This option can be specified multiple
times to get even more output.
.TP
\-x, --debug
Switches ib2roce into debug mode. ib2roce does not daemonize but
keeps logging diagnostic output to stderr. It will also present a command
line interface. See
.I bcom
for a description of available commands.
.TP
\-y, --disable <option>
Disable an optional features in the ib2roce daemon. See the section on optional features for details
.SH OPTIONAL FEATURES
.TP
.B bridging
Forwarding of packets by the bridge. This may be disabled for testing purposes to see if a configuration
brings up the bridge correctly and for inspection of data that may be flowing through the bridge if
it actually would start forwarding packets. Bridging is on by default.
.TP
.B drop
Enable dropping of packets (therefore packet loss!) for testing purposes. This is used to verify that
the recovery mechanisms work and that potential packet loss for other reasons will trigger proper
retransmit behavior by the applications using the ib2roce bridge. A number may be specified as a parameter which
indicates how many packets should be send before a drop ocurres. I.e.
.I enable drop=100
means that the 100th packet will not be forwarded. Drop is off by default.
.Tp
.B hwrate
Set the hardware forwarding rate of the NIC to a certain value. This is set in terms of the enum ibv_rate in ibverbs.h.
2= 2.5GBS, 3= 5GBGS and so on. NIC restrictions exist depending on the hardware vendor and fabric type that may cause this option to have no effect.
hwrate limitations are off by default.
.TP
.B loopbackprev
Configure the NIC to do loopback prevention. This is on by default and safe even for local apps when using the NVIDIA NICs. The implementation
of loopback prevention may vary for other vendors. Local multicast may not work if this option is on for certain RDMA NICs or software emulations.
For example the RXE driver is known to require this option.
.TP
.B ppsdisplay
Display a summary of the current traffic flow going through ib2roce bridge in intervals of
.I statint.
This option is off by default.
.TP
.B latency
Collect latency statistics for busy loop queues. This is used to verify the proper operation of the low latency cores and monitor the delays incurred by ib2roce as well as system processing. Results are visible in the "cores" output using the
.B bcom
tool. Off by default.
.TP
.B statint
Sampling interval for statistics and set to 10 seconds by default. This can be reconfigured to a different time interval or switches completely off.
.TP
.B loglevel
Logging level. Defaults to 6. Loglevel may be reduced to avoid informational messages or increased to get debugging messages.
.TP
.B buffers
The number of buffers to make available for ib2roce. Buffers are used for the send and receive queues and may be used to avoid the occasional hiccup on the fabrics.
Defaults to 200000 which is usually more than enough.
.TP
.B huge
Enable the use of huge pages for the buffers. This will significantly reduce TLB pressure and be useful if a large number of buffers is in use. The operating system needs
to be configured to provide a sufficient number of buffers otherwise ib2roce will fail to startup.
The default is off.
.TP
.B mcqp
The number of multicast groups to use for a QP (RDMA communication channel). The default is to use the maximum number of MC groups allowed by the NIC for each RDMA channel.
This minimizes the number of RDMA channels that the adapter nees to operate. The number may be set lower for testing purposes to see if ib2roce works correctly with a couple of rdma channels. Note that the maximum number of RDMA Channels supported per interface is limited to 10. One RDMA channel typically allows the use of 200 or more multicast groups. The maximum number of multcast groups on a fabric is limited at 2k or so. So this will be enough for the maximum number of multicast groups possible.
.TP
.B pgm
Enable PGM processing. This will track the PGM streams going through the bridge and allow diagnostics via the TSI bcom command. This is on by default.
.TP
.B pam_nak
Enable PGM NAK processing. Requires PGM processing to be on. ib2roce will redirect NAKs back to itselt. This avoids having to do unicast packet forwarding for NAKs separately via a different protocol (we can only route IP so the NAKs can then not use RDMA) and allows ib2roce to track NAKs instead of only the NCFs. This is an experimental feature and it is off by default.
.TP
.B sendrate
The rate of sending packets if ib2roce is requested to create a PGM stream for diagnostic purposes. See
.I mcsender.
The default is 5 packets per second.
.TP
.B sendlen
The size of the packet to create when ib2roce is requested to create a PGM stream for diagnostic purposes. See
.I mcsender.
The default is to only use the minimal packet size needed for the metadata which is usually less than 100 bytes. This cannot be set larger than the MTU.

.SH "NOTES"
ib2roce creates files in either the current directory or in /var/lib/ib2roce. These
are the ib2roce.pid file to be able to detect the presence of an already running
ib2roce instance and the ib2roce-status file that is updated in regular
intervals with ib2roce information.
.SH "SEE ALSO"
ib2roce(7), bcom(1), mclisten(1), mcsender(1)
.SH AUTHORS
Christoph Lameter <cl@linux.com>


ib2roce/ib2roce.c
/*
 * RDMA Infiniband to ROCE Bridge or Gateway
 *
 * (C) 2021-2022 Christoph Lameter <cl@linux.com>
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * $Author: Christoph Lameter [cl@linux.com]$
 *
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdarg.h>
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <ctype.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <numa.h>
#include <systemd/sd-daemon.h>

#include "errno.h"
#include "fifo.h"
#include "ring.h"
#include "hash.h"
#include "sched.h"
#include "logging.h"
#include "locking.h"
#include "buffers.h"
#include "multicast.h"
#include "interfaces.h"
#include "cli.h"
#include "pgm.h"
#include "sender.h"

#include "endpoint.h"

/* Globals */

static bool debug = false;		/* Stay in foreground, print more details */
static bool testing = false;		/* Run some tests on startup */
static uint64_t watchdog_nsec;
static int drop_packets = 0;		/* Packet dropper */

static char *payload_dump(uint8_t *p)
{
	return _hexbytes(p, 48);
}

static char *grh_str(struct ibv_grh *g)
{
        struct iphdr *i = (void *)g + 20;
        char xbuf[INET6_ADDRSTRLEN];
        char xbuf2[INET6_ADDRSTRLEN];
        char hbuf1[30];
        char hbuf2[30];
        struct in_addr saddr, daddr;
        static char buf[200];

        saddr.s_addr = i->saddr;
        daddr.s_addr = i->daddr;

        strcpy(hbuf1, inet_ntoa(saddr));
        strcpy(hbuf2, inet_ntoa(daddr));

        sprintf(buf, "GRH(flow=%u Len=%u next_hdr=%u hop_limit=%u SGID=%s DGID:%s SourceIP=%s DestIP=%s)",
                        ntohl(g->version_tclass_flow), ntohs(g->paylen), g->next_hdr, g->hop_limit,
                        inet_ntop(AF_INET6, &g->sgid, xbuf2, INET6_ADDRSTRLEN),
                        inet_ntop(AF_INET6, &g->dgid, xbuf, INET6_ADDRSTRLEN),
                        hbuf1, hbuf2);
	return buf;
}

/*
 * We have an GRH header so the packet has been processed by the RDMA
 * Subsystem and we can take care of it using the RDMA calls
 */
void receive_multicast(struct buf *buf)
{
	struct mc *m;
	struct rdma_channel *c = buf->c;
	enum interfaces in = c->i - i2r;
	struct ib_addr *dgid = (struct ib_addr *)&buf->grh.dgid.raw;
	struct in_addr dest_addr;
	int ret;
	const char *reason = NULL;

	if (!buf->grh_valid) {
		logg(LOG_WARNING, "No GRH on %s. Packet discarded: %s\n",
			c->text, payload_dump(buf->cur));
		goto invalid_packet;
	}

	if (buf->ip_valid) {

		if (!IN_MULTICAST(ntohl(buf->ip.daddr))) {
			reason = "Unicast Packet";
			goto discardit;
		}

		if (buf->ip.saddr == c->i->if_addr.sin_addr.s_addr && buf->w->qp_num == buf->w->src_qp) {
			reason = "Loopback Packet";
			goto discardit;
		}

		
		if (!__valid_addr(c->i, buf->ip.saddr)) {
			reason = "Packet not originating on source interface";
			goto discardit;
		}

	} else /* ! buf->ip_valid */ {

		if (buf->grh.dgid.raw[0] !=  0xff) {
			reason = "Unicast Packet";
			goto discardit;
		}

		if (memcmp(&buf->grh.sgid, &c->i->gid, sizeof(union ibv_gid)) == 0 && buf->w->qp_num == buf->w->src_qp) {

			reason = "Loopback Packet";
			goto discardit;
		}

		/*
		 * ib2roce sets the hop limit to 1. By default is it
		 * 0. So if hop_limit is set then another ib2roce already
		 * processed the packet. Discard it.
		 */
		if (buf->grh.hop_limit) {
			reason = "Hop Limit !=0 discard packet.\n";
			goto discardit;
		}
	}

	dest_addr.s_addr = dgid->sib_addr32[3];
	m = hash_lookup_mc(dest_addr);

	logg(LOG_DEBUG, "From %s: MC=%s\n", c->text, inet_ntoa(dest_addr));

	if (!m) {
		logg(LOG_INFO, "Discard Packet: Multicast group %s not found\n",
			inet_ntoa(dest_addr));
		goto invalid_packet;
	}

	if (m->interface[in].sendonly) {

		logg(LOG_INFO, "Discard Packet: Received data from Sendonly MC group %s from %s\n",
			m->text, c->text);
		goto invalid_packet;
	}

	if (!buf->ip_valid) {
		unsigned char *mgid = buf->grh.dgid.raw;
		unsigned short signature = ntohs(*(unsigned short*)(mgid + 2));

		if (!mgid_check(m, signature)) {
			logg(LOG_INFO, "Discard Packet: MGID multicast signature(%x)  mismatch. MGID=%s\n",
					signature, inet6_ntoa(mgid));
			goto invalid_packet;
		}

	}

	if (m->callback) {
		m->callback(m, in, buf);
		return;
	}

	if (!m->enabled)
		return;

	if (pgm_mode && m->mgid_mode == MGID_MODE_PORT) {
		uint8_t *saved = buf->cur;
		if (!pgm_process(c, m, buf))
			return;
		buf->cur = saved;
	}

	if (!bridging)
		return;

	if (drop_packets && (c->stats[packets_received] % drop_packets) == drop_packets - 1)
		return;

	struct mc_interface *mi = m->interface  + (in ^ 1);
	struct rdma_channel *ch_out = mi->channel;

	if (!m->same_core) {
		/* Ok we need to queue on another core */
		fifo_put(&ch_out->send_queue, buf);
		goto success;
	}

	if (!mi->ai.ah)		/* After a join it may take awhile for the ah pointer to propagate */
 		sleep(1);

	get_buf(buf);	/* Packet will not be freed on return from this function */
 
	ret = send_to(ch_out, buf->cur, buf->end - buf->cur, &mi->ai, buf->imm_valid, buf->imm, buf);
 	if (ret)
		return;

success:
	st(c, packets_bridged);
	return;

discardit:    
	logg(LOG_DEBUG, "%s on multicast channel %s: GRH=%s %s\n", reason, c->text, grh_str(&buf->grh), payload_dump(buf->cur));

invalid_packet:
	st(c, packets_invalid);
}

void receive_unicast(struct buf *buf)
{
	struct rdma_channel *c = buf->c;

	logg(LOG_DEBUG, "Drop on unicast channel %s: %s\n", c->text, payload_dump(buf->cur));
	st(c, packets_invalid);
}

static void run_watchdog(void *private)
{
	sd_notify(0, "WATCHDOG=1");
	add_event(timestamp() + watchdog_nsec, run_watchdog, NULL, "Watchdog");
}

static void setup_timed_events(void)
{
	uint64_t x;
	int ret;

	now = timestamp();

	start_calculate_pps();

	check_joins(&i2r[INFINIBAND].channels, &i2r[ROCE].channels);

	ret = sd_watchdog_enabled(0, &x);
	if (ret < 0)
		panic("Watchdog check failed. errname())\n");

	if (!ret)
		return;

	if (x < 100000)
		panic("Watchdog timer less than 100 milliseconds not supported\n");

	if (!systemd)
		panic("Watchdog only supported in --systemd mode\n");

	/* Systemd recommends to send watchdog notifications in half the requested time interval */
	watchdog_nsec = x * ONE_MICROSECOND /2;
	run_watchdog(NULL);
}

static void setup_termination_signals(void)
{
	signal(SIGINT, terminate);
	signal(SIGTERM, terminate);
	signal(SIGHUP, terminate);	/* Future: Reload a potential config file */
}

static void daemonize(void)
{
	pid_t pid;

	pid = fork();

	if (pid < 0)
		exit(EXIT_FAILURE);

	if (pid > 0)
		exit(EXIT_SUCCESS);

	if (setsid() < 0)
		exit(EXIT_FAILURE);

	signal(SIGCHLD, SIG_IGN);
	signal(SIGHUP, SIG_IGN);

	pid = fork();

	if (pid < 0)
		exit(EXIT_FAILURE);

	/* Terminate parent */
	if (pid > 0)
		exit(EXIT_SUCCESS);

	/* Set new file permissions */
	umask(0);

	if (chdir("/var/lib/ib2roce")) {
		perror("chdir");
		printf("/var/lib/ib2roce must exist and be writeable for daemon mode.\n");
		exit(EXIT_FAILURE);
	}

	/* Close all open file descriptors */
	int x;
	for (x = sysconf(_SC_OPEN_MAX); x>2; x--)
		close(x);
	close(0);

	openlog ("ib2roce", LOG_PID, LOG_DAEMON);
}

static int pid_fd;

static char pid_name[40];

static void pid_open(void)
{
	struct flock fl = {
		.l_type = F_WRLCK,
		.l_whence = SEEK_SET,
		.l_start = 0,
		.l_len = 0
	};
	int n;
	char buf[10];

	if (default_port)
		snprintf(pid_name, sizeof(pid_name), "ib2roce-%d.pid", default_port);
	else
		strcpy(pid_name, "ib2roce.pid");

	pid_fd = open(pid_name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);

	if (pid_fd < 0)
		panic("Cannot open pidfile. Error %s\n", errname());

	if (fcntl(pid_fd, F_SETLK, &fl) < 0)
       		panic("ib2roce already running.\n");

	if (ftruncate(pid_fd, 0) < 0)
       		panic("Cannot truncate pidfile. Error %s\n", errname());

	n = snprintf(buf, sizeof(buf), "%ld", (long) getpid());

	if (write(pid_fd, buf, n) != n)
       		panic("Cannot write pidfile. Error %s\n", errname());
}

static void pid_close(void)
{
	unlink(pid_name);
	close(pid_fd);
}

static void debug_set(char *optarg)
{
	debug = true;
}

static void test_set(char *optarg)
{
	ring_test();
	fifo_test();
	hash_test();
	testing = true;
}

static void setup_options(void)
{
	register_option("debug", no_argument, 'x', debug_set,
		       	NULL, "Do not daemonize, enter command line mode");
	register_option("test", no_argument, 'q', test_set, NULL, "Run selftest");
}

static void setup_enable(void)
{
	register_enable("bridging", false, &bridging, NULL, "on", "off", NULL,
		"Forwarding of packets between interfaces");
	register_enable("drop", true, NULL, &drop_packets, "100", "0", NULL,
		"Drop multicast packets. The value is the number of multicast packets to send before dropping");
	register_enable("hwrate", true, NULL, &rate, "6", "0", NULL,
		"Set the speed in the RDMA NIC to limit the output speed 2 =2.5GBPS 5 = 5GBPS 3 = 10GBPS ...(see enum ibv_rate)");
}

__attribute__((constructor))
static void options_init(void)
{
	setup_options();
	setup_enable();
}

int main(int argc, char **argv)
{
	int ret = 0;

	/* Init the scheduler */
	now = timestamp();
	parse_options(argc, argv);

	if (debug || !bridging) {

		openlog("ib2roce", LOG_PERROR, LOG_USER);

	} else {
		background = true;
		if (!systemd) {
			daemonize();
			pid_open();
		}
	}
	concom_init();

	setup_termination_signals();

	ret = find_rdma_devices();
	if (ret && !testing)
		return ret;

	if (mode == mode_bridge)
		logg (LOG_NOTICE, "%s device = %s:%d, %s device = %s:%d. Multicast Groups=%d MGIDs=%s Buffers=%u\n",
			interfaces_text[INFINIBAND], i2r[INFINIBAND].rdma_name, i2r[INFINIBAND].port,
			interfaces_text[ROCE], i2r[ROCE].rdma_name, i2r[ROCE].port,
			nr_mc, mgid_text(NULL), nr_buffers);
	else {
		logg (LOG_NOTICE, "%s device = %s:%d. Multicast Groups=%d MGIDs=%s Buffers=%u\n",
			interfaces_text[default_interface], i2r[default_interface].rdma_name, i2r[default_interface].port,
			nr_mc, mgid_text(NULL), nr_buffers);
	}

	numa_run_on_node(i2r[INFINIBAND].context ? i2r[INFINIBAND].numa_node : i2r[ROCE].numa_node);
	init_buf();	/* Setup interface registers memmory */
	numa_run_on_node(-1);

	setup_interface(INFINIBAND);
	setup_interface(ROCE);

	if (!i2r[INFINIBAND].context && !i2r[ROCE].context) {
		logg(LOG_CRIT, "No working RDMA devices present.\n");
		exit(2);
	}

	if (cores)
		show_core_config();

	if (mode == mode_sender)
		sender_setup();

	post_receive_buffers();

	start_cores();
	arm_channels(NULL);
	setup_timed_events();

	if (systemd)
		sd_notify(0, "READY=1");

	if (event_loop() < 0)
		logg(LOG_ERR, "Event Loop failed with %s\n", errname());

	if (systemd)
		sd_notify(0, "STOPPING=1");

	if (mode == mode_sender)
		sender_shutdown();

	stop_cores();

	shutdown_roce();
	shutdown_ib();

	if (background)
		pid_close();

	syslog (LOG_NOTICE, "Shutdown complete.\n");
	closelog();

	return EXIT_SUCCESS;
}
